File: backend/ambi/services/pinecone_memory_service.py
"""Pinecone-based memory service for Ambi Python backend."""

import logging
import uuid
from typing import Dict, List, Optional

from langchain.embeddings import OpenAIEmbeddings
from langchain.vectorstores import Pinecone as LangchainPinecone

from ambi.config.settings import PINECONE_INDEX_NAME
from ambi.db.pinecone import pinecone_connection

logger = logging.getLogger(__name__)


class PineconeMemoryService:
    """
    Pinecone-based memory service for long-term semantic memory.
    
    This service uses Pinecone vector database to store and retrieve
    memories based on semantic similarity.
    """

    def __init__(self) -> None:
        """Initialize the Pinecone memory service."""
        self.pinecone = pinecone_connection.get_index()
        self.embeddings = OpenAIEmbeddings()

    async def add_memory(
        self,
        session_id: str,
        text: str,
        metadata: Optional[Dict] = None,
    ) -> bool:
        """
        Add a memory entry to the long-term memory.
        
        Args:
            session_id: Unique identifier for the conversation session
            text: Text content to store
            metadata: Optional metadata for the memory entry
            
        Returns:
            Success status of the operation
        """
        if not self.pinecone:
            logger.error("Pinecone index is not initialized")
            return False

        if not session_id or not text:
            logger.error("Session ID and text are required")
            return False

        try:
            if metadata is None:
                metadata = {}
            
            metadata["session_id"] = session_id
            
            memory_id = f"{session_id}:{uuid.uuid4()}"
            
            embedding = await self._get_embedding(text)
            if not embedding:
                logger.error("Failed to generate embedding for text")
                return False
            
            self.pinecone.upsert(
                vectors=[
                    {
                        "id": memory_id,
                        "values": embedding,
                        "metadata": {
                            "text": text,
                            **metadata,
                        },
                    }
                ]
            )
            
            return True
            
        except Exception as e:
            logger.error(f"Error adding memory to Pinecone: {str(e)}")
            return False

    async def search_memories(
        self,
        session_id: str,
        query: str,
        limit: int = 5,
    ) -> List[Dict]:
        """
        Search for relevant memories based on semantic similarity.
        
        Args:
            session_id: Unique identifier for the conversation session
            query: Search query text
            limit: Maximum number of results to retrieve
            
        Returns:
            List of relevant memory entries
        """
        if not self.pinecone:
            logger.error("Pinecone index is not initialized")
            return []

        if not session_id or not query:
            logger.error("Session ID and query are required")
            return []

        try:
            embedding = await self._get_embedding(query)
            if not embedding:
                logger.error("Failed to generate embedding for query")
                return []
            
            search_response = self.pinecone.query(
                vector=embedding,
                filter={"session_id": session_id},
                top_k=limit,
                include_metadata=True,
            )
            
            results = []
            for match in search_response.get("matches", []):
                metadata = match.get("metadata", {})
                results.append({
                    "text": metadata.get("text", ""),
                    "score": match.get("score", 0.0),
                    "metadata": {k: v for k, v in metadata.items() if k != "text"},
                })
            
            return results
            
        except Exception as e:
            logger.error(f"Error searching memories in Pinecone: {str(e)}")
            return []

    async def clear_session(self, session_id: str) -> bool:
        """
        Clear all memories for a specific session.
        
        Args:
            session_id: Unique identifier for the conversation session
            
        Returns:
            Success status of the operation
        """
        if not self.pinecone:
            logger.error("Pinecone index is not initialized")
            return False

        if not session_id:
            logger.error("Session ID is required")
            return False

        try:
            self.pinecone.delete(
                filter={"session_id": session_id},
            )
            
            return True
            
        except Exception as e:
            logger.error(f"Error clearing session from Pinecone: {str(e)}")
            return False

    async def _get_embedding(self, text: str) -> Optional[List[float]]:
        """
        Generate embedding for text using the configured embedding model.
        
        Args:
            text: Text to generate embedding for
            
        Returns:
            Embedding vector as a list of floats
        """
        try:
            embedding = self.embeddings.embed_query(text)
            return embedding
            
        except Exception as e:
            logger.error(f"Error generating embedding: {str(e)}")
            return None


pinecone_memory_service = PineconeMemoryService()

File: backend/src/services/pineconeMemoryService.ts
import { BaseMessage, HumanMessage, AIMessage } from '@langchain/core/messages';
import { BaseMemory, InputValues, OutputValues, MemoryVariables } from '@langchain/core/memory';
import { PineconeRecord } from '@pinecone-database/pinecone';
import { v4 as uuidv4 } from 'uuid';
import { longTermMemoryConfig, categorizationConfig } from '../config/memoryConfig';
import { upsertVectors, queryVectors } from '../clients/pineconeClient';
import { generateEmbedding } from './embeddingService';

/**
 * Pinecone-backed implementation of long-term memory using vector embeddings.
 * This service provides semantic search and retrieval of past conversations.
 */

/**
 * Interface for memory entry metadata.
 */
interface _MemoryEntryMetadata {
  sessionId: string;
  timestamp: number;
  role: 'human' | 'ai';
  category?: string;
  priority?: number;
  tags?: string[];
}

/**
 * Custom implementation of a vector store memory that uses Pinecone for storage.
 * This class provides semantic search and retrieval of past conversations.
 */
export class PineconeVectorMemory extends BaseMemory {
  private namespace: string;
  private inputKey: string;
  private outputKey: string;
  private relevanceThreshold: number;
  private maxResults: number;
  
  constructor({
    namespace = longTermMemoryConfig.pineconeNamespace,
    inputKey = 'input',
    outputKey = 'output',
    relevanceThreshold = longTermMemoryConfig.relevanceThreshold,
    maxResults = longTermMemoryConfig.maxResults,
  } = {}) {
    super();
    this.namespace = namespace;
    this.inputKey = inputKey;
    this.outputKey = outputKey;
    this.relevanceThreshold = relevanceThreshold;
    this.maxResults = maxResults;
  }
  
  get memoryKeys(): string[] {
    return [this.inputKey, this.outputKey];
  }
  
  /**
   * Loads memory variables based on the input values.
   * @param values - The input values.
   * @returns A promise that resolves with the memory variables.
   */
  async loadMemoryVariables(values: InputValues): Promise<MemoryVariables> {
    const query = values[this.inputKey] as string;

    if (!query) {
      return { relevantHistory: [] };
    }

    try {
      const queryEmbedding = await generateEmbedding(query);

      // Increase topK to fetch more potential matches (human + AI pairs)
      const increasedTopK = this.maxResults * 2; // Example: fetch double

      const results = await queryVectors(
        queryEmbedding,
        increasedTopK,
        this.namespace
      );

      if (!results || !results.matches.length) {
        this.log("No matches found in Pinecone.");
        return { relevantHistory: [] };
      }

      this.log(`Initial Pinecone query returned ${results.matches.length} matches.`);

      // Define a type for matches with metadata and score
      type MatchWithMetadata = {
          id: string;
          score?: number;
          metadata?: Record<string, unknown>;
      }

      // Group matches by session ID
      const matchesBySession: Record<string, MatchWithMetadata[]> = {};
      results.matches.forEach(match => {
        const metadata = match.metadata as Record<string, unknown>; // Assuming metadata exists
        const sessionId = metadata?.sessionId as string;
        if (sessionId && match.id) { // Ensure match has id and sessionId
          if (!matchesBySession[sessionId]) {
            matchesBySession[sessionId] = [];
          }
          matchesBySession[sessionId].push({
              id: match.id,
              score: match.score,
              metadata: metadata 
          });
        }
      });

      const relevantMessages: BaseMessage[] = [];

      // Process each session separately to find pairs
      for (const sessionId in matchesBySession) {
        const sessionMatches = matchesBySession[sessionId];
        // Sort by timestamp to identify pairs
        sessionMatches.sort((a, b) => (a.metadata?.timestamp as number) - (b.metadata?.timestamp as number));

        for (let i = 0; i < sessionMatches.length; i++) {
          const match = sessionMatches[i];
          const metadata = match.metadata as Record<string, unknown>; 
          const score = match.score || 0;

          // If we find a relevant human message, look for the subsequent AI message
          if (metadata?.role === 'human' && score >= this.relevanceThreshold) {
            const humanContent = metadata?.originalContent as string; // Get content from metadata
            if (humanContent) { // Check if content exists in metadata
              relevantMessages.push(new HumanMessage(humanContent));
            }

            // Look for the next message in the sorted list IF it's an AI message
            if (i + 1 < sessionMatches.length) {
              const nextMatch = sessionMatches[i + 1];
              const nextMetadata = nextMatch.metadata as Record<string, unknown>; 
              if (nextMetadata?.role === 'ai') {
                const aiContent = nextMetadata?.originalContent as string; // Get content from metadata
                if (aiContent) { // Check if content exists in metadata
                  relevantMessages.push(new AIMessage(aiContent));
                }
              }
            }
          }
        }
      }
      
      // Deduplicate messages
      const uniqueMessages = relevantMessages.filter((msg, index, self) =>
        index === self.findIndex((m) => JSON.stringify(m) === JSON.stringify(msg))
      );

      this.log(`Retrieved ${uniqueMessages.length} relevant messages after pairing.`);
      return { relevantHistory: uniqueMessages }; 
    } catch (error) {
      console.error('[PineconeMemoryService] Error loading memory variables:', error);
      return { relevantHistory: [] };
    }
  }
  
  /**
   * Saves the context based on the input and output values.
   * @param inputValues - The input values.
   * @param outputValues - The output values.
   * @returns A promise that resolves when the context has been saved.
   */
  async saveContext(inputValues: InputValues, outputValues: OutputValues): Promise<void> {
    const humanContent = inputValues[this.inputKey] as string;
    const aiContent = outputValues[this.outputKey] as string;
    
    if (!humanContent || !aiContent) {
      return;
    }
    
    try {
      const [humanEmbedding, aiEmbedding] = await Promise.all([
        generateEmbedding(humanContent),
        generateEmbedding(aiContent)
      ]);
      
      const sessionId = inputValues.sessionId as string || uuidv4();
      
      const humanCategory = categorizeMessage(humanContent);
      const aiCategory = categorizeMessage(aiContent);
      
      const humanRecord: PineconeRecord = {
        id: `${sessionId}:human:${uuidv4()}`, // Use UUID instead of content for ID
        values: humanEmbedding,
        metadata: {
          sessionId,
          timestamp: Date.now(),
          role: 'human',
          category: humanCategory,
          priority: getPriorityLevel(humanContent),
          originalContent: humanContent, // Store original content in metadata
        },
      };
      
      const aiRecord: PineconeRecord = {
        id: `${sessionId}:ai:${uuidv4()}`, // Use UUID instead of content for ID
        values: aiEmbedding,
        metadata: {
          sessionId,
          timestamp: Date.now(),
          role: 'ai',
          category: aiCategory,
          priority: getPriorityLevel(aiContent),
          originalContent: aiContent, // Store original content in metadata
        },
      };
      
      await upsertVectors([humanRecord, aiRecord], this.namespace);
      console.log(`[PineconeMemoryService] Saved conversation to long-term memory.`);
    } catch (error) {
      console.error('[PineconeMemoryService] Error saving context:', error);
    }
  }
  
  /**
   * Clears all memories from the vector store.
   * @returns A promise that resolves when the memories have been cleared.
   */
  async clear(): Promise<void> {
    try {
      const { deleteAllVectors } = await import('../clients/pineconeClient');
      await deleteAllVectors(this.namespace);
      console.log(`[PineconeMemoryService] Cleared all memories from namespace ${this.namespace}.`);
    } catch (error) {
      console.error(`[PineconeMemoryService] Error clearing memories:`, error);
    }
  }

  // Helper log function
  private log(message: string) {
    console.log(`[PineconeMemoryService] ${message}`);
  }
}

/**
 * Creates a Pinecone-backed vector memory for semantic search and retrieval.
 * @param sessionId - Optional session ID to associate with the memory.
 * @returns A PineconeVectorMemory instance.
 */
export const createPineconeMemory = (_sessionId?: string): PineconeVectorMemory => {
  return new PineconeVectorMemory();
};

/**
 * Simple categorization function for messages.
 * In a real implementation, this would use more sophisticated NLP techniques.
 * @param content - The message content to categorize.
 * @returns The category of the message.
 */
const categorizeMessage = (content: string): string => {
  const lowerContent = content.toLowerCase();
  const categories = categorizationConfig.categories;
  
  if (lowerContent.includes('name') || lowerContent.includes('call me') || lowerContent.includes('my name is')) {
    return categories[0]; // personal_info
  }
  
  if (lowerContent.includes('like') || lowerContent.includes('prefer') || lowerContent.includes('favorite')) {
    return categories[1]; // preferences
  }
  
  if (lowerContent.includes('family') || lowerContent.includes('parent') || 
      lowerContent.includes('child') || lowerContent.includes('sister') || 
      lowerContent.includes('brother') || lowerContent.includes('spouse')) {
    return categories[2]; // family
  }
  
  if (lowerContent.includes('health') || lowerContent.includes('sick') || 
      lowerContent.includes('doctor') || lowerContent.includes('pain')) {
    return categories[3]; // health
  }
  
  if (lowerContent.includes('event') || lowerContent.includes('meeting') || 
      lowerContent.includes('appointment') || lowerContent.includes('schedule')) {
    return categories[4]; // events
  }
  
  return categories[5]; // general
};

/**
 * Determines the priority level of a message based on its content.
 * @param content - The message content to analyze.
 * @returns The priority level (1-3).
 */
const getPriorityLevel = (content: string): number => {
  const lowerContent = content.toLowerCase();
  const { priorityLevels } = categorizationConfig;
  
  if (lowerContent.includes('name') || lowerContent.includes('address') || 
      lowerContent.includes('phone') || lowerContent.includes('email') ||
      lowerContent.includes('birthday') || lowerContent.includes('important')) {
    return priorityLevels.high;
  }
  
  if (lowerContent.includes('like') || lowerContent.includes('prefer') || 
      lowerContent.includes('event') || lowerContent.includes('meeting') ||
      lowerContent.includes('remember') || lowerContent.includes('don\'t forget')) {
    return priorityLevels.medium;
  }
  
  return priorityLevels.low;
};
